### PR TITLE
feat: Implement encryption when exchanging challenge

### PR DESCRIPTION
**Pull request by didrocks**
This generates an asymmetric private/public key pair at broker creation time. We let it panic as this broker is only for testing purposes.
This encryption key is then passed at session creation. Note that the broker can create a new key for each session if desired.

The chosen encryption is sha512. The public key is passed as X.509 standard, encoded as base64.

Testing is only covered at integration test level, as this requires a broker and exercise the PAM module itself.

UDENG-2014

### CODE REVIEW

**Review by jibel — APPROVED** [review]
Fine for me. Thanks!

File: internal/brokers/examplebroker/broker.go [hunk 1]
```diff
@@ -70,19 +77,22 @@ var (
 	}
 )
 
-const (
-	brokerEncryptionKey = "encryptionkey"
-)
-
 // New creates a new examplebroker object.
 func New(name string) (b *Broker, fullName, brandIcon string) {
+	// Generate a new private/public keys for this session.
```
Comment (denisonbarbosa) [inline internal/brokers/examplebroker/broker.go:82]:
```suggestion
	// Generate a new private key for the broker.
```

**Review by denisonbarbosa — APPROVED** [review]
Looks great! Just a minor comment fix and we can merge.

File: pam/internal/adapter/model.go [hunk 2]
```diff
@@ -201,7 +228,7 @@ func (m *UIModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
 		log.Info(context.TODO(), "UILayoutReceived")
 
 		return m, tea.Sequence(
-			m.authenticationModel.Compose(m.currentSession.brokerID, m.currentSession.sessionID, msg.layout),
+			m.authenticationModel.Compose(m.currentSession.brokerID, m.currentSession.sessionID, m.currentSession.encryptionKey, msg.layout),
```
Comment (3v1n0) [inline pam/internal/adapter/model.go:231]: > 120 chars wide :)

**Comment by 3v1n0** [comment]
While I see the point of sending encrypted data to the broker (even though they should definitely be running in a system bus, so if someone can intercept the traffic we're screwed anyways)... Do we need to have this handled by the pam module?

We're again doing communication on a private socket, so I feel we could move the encryption side on the broker manager part, instead of handling it in the two very ends of the wire, no?

File: pam/internal/adapter/model.go (hunk 2 above)
Comment (didrocks) [reply pam/internal/adapter/model.go:231]: Not our limit :p (as already argued in other PRs), here we have 2 sequences and it reads better that way

**Comment by didrocks** [comment]
@3v1n0 when we discussed with the sssd team about the protocol when working on the common projects, they really wanted to have this encrypted from end to end, as far as we can, I know the gdm side is not doing this and maybe this is something to fix there in the long term, but at least, from a protocol standpoint, we did our best to encrypt as largely as possible. So, we are going to keep the encryption as such and revisit if the security team has any concern.

File: pam/internal/adapter/model.go (hunk 2 above)
Comment (3v1n0) [reply pam/internal/adapter/model.go:231]: So we said, didn't we? :)

File: internal/brokers/examplebroker/broker.go (hunk 1 above)
Comment (didrocks) [reply internal/brokers/examplebroker/broker.go]: Fixed!

**Comment by 3v1n0** [comment]
> I know the gdm side is not doing this and maybe this is something to fix there in the long term, but at least, from a protocol standpoint, we did our best to encrypt as largely as possible. So, we are going to keep the encryption as such and revisit if the security team has any concern.

I see... We may still handle this in gdm too now, over the JSON protocol, not sure if it's worth it (as again if someone gets control of whats transmitted in the gdm dbus bus), then... We're screwed.

File: pam/internal/adapter/model.go (hunk 2 above)
Comment (didrocks) [reply pam/internal/adapter/model.go:231]: I don't think we never acknowledged to it, or if so, we would have a checker. We have lines of 130 characters in other projects and it seems that didn’t create any issue inside desktop enterprise/wsl team.

So I suggest either we all discuss this to have it as a rule in our standup, and if so, write a linter, ensure that all projects pass to it, or just deal with it :) (for now, I would argue that the line is readalable and wrapping make it harder to parse the sequence itself)

**Pull request merged**